File: Projet Poo/Joueurs/Suggestion.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using LibCLR;

namespace Modele
{
    public class Suggestion : StrategySugg
    {
        protected enum PeupleInt { Gaulois = 1, Viking = 0, Nain = 2 };
        protected enum CaseInt { Plaine = 0, Eau = 1, Montagne = 2, Desert = 3, Foret = 4 };

        public override SuggMap getSuggestion(Carte carte, Unite unite)
        {

            int x = unite.Coord.X;
            int y = unite.Coord.Y;
            WrapperSuggestion wrap = new WrapperSuggestion();

            List<int> emplUnites = getUnitesToList(carte, unite);
            List<int> carteInt = getCasesToList(carte);

            PeupleInt peuple = PeupleInt.Gaulois;
            IPeuple p = unite.Proprietaire.Peuple;

            if (p is PeupleViking)
                peuple = PeupleInt.Viking;
            else if (p is PeupleNain)
                peuple = PeupleInt.Nain;
            else if (p is PeupleGaulois)
                peuple = PeupleInt.Gaulois;

            List<int> sugg = wrap.getSuggestion(carteInt, emplUnites, carte.Largeur, x, y, unite.PointsDepl, (int)peuple);

            SuggMap res = new SuggMap();
            for (int i = 0; i < carte.Largeur; i++)
            {
                for (int j = 0; j < carte.Hauteur; j++)
                {
                    SuggData data = new SuggData();
                    data.Sugg = sugg[i * carte.Largeur + j];
                    res[new Coordonnees(i, j)] = data;
                }
            }

            for (int i = 0; i < carte.Largeur; i++)
            {
                for (int j = 0; j < carte.Hauteur; j++)
                {
                    res[new Coordonnees(i, j)].Depl = sugg[carte.Largeur * carte.Hauteur + (i * carte.Largeur + j)];
                }
            }

            return res;
        }

        protected List<int> getCasesToList(Carte carte)
        {
            List<int> res = new List<int>();
            for (int i = 0; i < carte.Largeur; i++)
            {
                for (int j = 0; j < carte.Hauteur; j++)
                {
                    CaseInt tile = CaseInt.Plaine;
                    if (carte.Cases[i][j] is CaseDesert)
                    {
                        tile = CaseInt.Desert;
                    }
                    else if (carte.Cases[i][j] is CaseEau)
                    {
                        tile = CaseInt.Eau;
                    }
                    else if (carte.Cases[i][j] is CaseForet)
                    {
                        tile = CaseInt.Foret;
                    }
                    else if (carte.Cases[i][j] is CaseMontagne)
                    {
                        tile = CaseInt.Montagne;
                    }
                    else if (carte.Cases[i][j] is CasePlaine)
                    {
                        tile = CaseInt.Plaine;
                    }
                    res.Add((int)tile);
                }
            }
            return res;
        }

        protected List<int> getUnitesToList(Carte carte, Unite unite)
        {
            List<int> res = new List<int>();
            List<Unite> tmp;
            for (int i = 0; i < carte.Largeur; i++)
            {
                for (int j = 0; j < carte.Hauteur; j++)
                {
                    tmp = carte.getUniteFromCoord(new Coordonnees(i, j));
                    if (tmp.Count > 0 && tmp[0].IdProprietaire != unite.IdProprietaire)
                        res.Add(tmp.Count);
                    else
                        res.Add(0);
                }
            }

            return res;
        }
    }
}

File: LibCpp/Suggestion.cpp
#include "Suggestion.h"


Suggestion::Suggestion(void)
{
}


Suggestion::~Suggestion(void)
{
}


vector<int*> Suggestion::suggestion(int **carte, int**unites, int largeur, int hauteur,int xActuel,int yActuel, int ptDepl, int peupleJActuel)
{
	 
	int *** sugg = new int**[largeur];
	for (int i=0; i<largeur; i++) 
	{
		sugg[i]= new int*[hauteur];
		for (int j=0; j<hauteur; j++) {
			sugg[i][j] = new int[2];
			sugg[i][j][0] = 0;
			sugg[i][j][1] = 0;
		} 
	}
	
	sugg[xActuel][yActuel][0]=verif[peupleJActuel][carte[xActuel][yActuel]];
	calculDeplClassique(carte,unites,xActuel,yActuel,ptDepl,peupleJActuel,largeur,hauteur, sugg);
	vector<int*> res;
	int *data;
	for (int i=0; i<largeur; i++) 
	{
		for (int j=0; j<hauteur; j++) {
			data = new int[2];
			data[0]=sugg[i][j][0];
			data[1]=sugg[i][j][1];
			res.push_back(data);
		} 
	}

	return res;
}

 
void Suggestion::calculDeplClassique(int **carte, int **unites, int xActuel, int yActuel, int depl, int peupleJActuel, int largeur, int hauteur, int*** sugg)
{
	if (depl>0)
	{
		sugg[xActuel][yActuel][1]=depl;
		int deplSuiv;
		if (peupleJActuel==Elfe && carte[xActuel][yActuel]==Foret)
			deplSuiv=depl;
		else
			deplSuiv=depl-1;

		if (xActuel-1>=0)
		{
			if (unites[xActuel][yActuel]==0)
			{
				sugg[xActuel-1][yActuel][0]=verif[peupleJActuel][carte[xActuel-1][yActuel]];
				if (sugg[xActuel-1][yActuel][1]<depl && sugg[xActuel-1][yActuel][0]>0)
					calculDeplClassique(carte,unites,xActuel-1,yActuel,deplSuiv,peupleJActuel,largeur,hauteur,sugg);
			}
		}
		if (xActuel+1<largeur)
		{
			if (unites[xActuel][yActuel]==0)
			{
				sugg[xActuel+1][yActuel][0]=verif[peupleJActuel][carte[xActuel+1][yActuel]];
				if (sugg[xActuel+1][yActuel][1]<depl && sugg[xActuel+1][yActuel][0]>0)
					calculDeplClassique(carte,unites,xActuel+1,yActuel,deplSuiv,peupleJActuel,largeur,hauteur,sugg);
			}
		}
		if (yActuel-1>=0)
		{
			if (unites[xActuel][yActuel]==0)
			{
				sugg[xActuel][yActuel-1][0]=verif[peupleJActuel][carte[xActuel][yActuel-1]];
				if (sugg[xActuel][yActuel-1][1]<depl && sugg[xActuel][yActuel-1][0]>0)
					calculDeplClassique(carte,unites,xActuel,yActuel-1,deplSuiv,peupleJActuel,largeur,hauteur,sugg);
			}
		}
		if (yActuel+1<hauteur)
		{
			if (unites[xActuel][yActuel]==0)
			{
				sugg[xActuel][yActuel+1][0]=verif[peupleJActuel][carte[xActuel][yActuel+1]];
				if (sugg[xActuel][yActuel+1][1]<depl && sugg[xActuel][yActuel+1][0]>0)
					calculDeplClassique(carte,unites,xActuel,yActuel+1,deplSuiv,peupleJActuel,largeur,hauteur,sugg);
			}
		}

		if (peupleJActuel==Nain && carte[xActuel][yActuel]==Montagne)
        {
            for (int i=0;i<largeur;i++)
            {
                for (int j=0;j<hauteur;j++)
                {
                    if(carte[i][j]==Montagne && unites[i][j]==0)
                    {
                        sugg[i][j][0]=2;
                        if (sugg[i][j][1]<depl && sugg[i][j][0]>0)
                            calculDeplClassique(carte,unites,i,j,depl-1,peupleJActuel,largeur,hauteur, sugg);
                    }
                }
            }
        }
		else if(peupleJActuel==Viking)
		{
			if (xActuel-1>=0)
			{
				if (carte[xActuel-1][yActuel]==Eau)
					sugg[xActuel][yActuel][0]=5;
			}
			if (xActuel+1<largeur)
			{
				if (carte[xActuel+1][yActuel]==Eau)
					sugg[xActuel][yActuel][0]=5;
			}
			if (yActuel-1>=0)
			{
				if (carte[xActuel][yActuel-1]==Eau)
					sugg[xActuel][yActuel][0]=5;
			}
			if (yActuel+1<hauteur)
			{
				if (carte[xActuel][yActuel+1]==Eau)
					sugg[xActuel][yActuel][0]=5;
			}
		}
	}
}